libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -233,6 +234,15 @@ emptyGenState proof genv =
     proof
     genv
     0
+  where
+    f :: Reflect era => Proof era -> GenEnv era -> ModelNewEpochState era
+    f proof' genv' = case proof' of
+      Shelley _ -> mNewEpochStateZero {mPParams = gePParams genv'}
+      Mary _ -> mNewEpochStateZero {mPParams = gePParams genv'}
+      Allegra _ -> mNewEpochStateZero {mPParams = gePParams genv'}
+      Alonzo _ -> mNewEpochStateZero {mPParams = gePParams genv'}
+      Babbage _ -> mNewEpochStateZero {mPParams = gePParams genv'}
+      Conway _ -> mNewEpochStateZero {mPParams = gePParams genv'}
```

Comment:
It's necessary so that we don't add the `EraTallyState era` constraint to the function. I could push the pattern match into `mNewEpochStateZero`, but it'll have to happen at some point since we're calling `emptyTallyState`.

Reply:
There is nothing wrong with adding `EraTallyState` constraint to the function. It is a type class that is satisfied for all eras. Same applies to all of the classes: `Era`, `EraTxBody`, `EraTxOut`, etc. but not to the era specific ones. So adding `AlonzoEraTxOut` would not work.

Reply:
I think the idea behind the `Generic` modules is that we try to have as few constraints as possible in the function signatures.

Reply:
That's not the idea! Adding a constraints that is always satisfiable should never be a problem.

The idea of Generic tests is to be able to do different things that are not provided by any of the type classes and are specific to each era. In this case you don't need to rely on `proof` and can do it by relying on the `EraTallyState` constraint, so you should take this option instead, because it reduces boiler plate and is a lot less verbose.

If we were to follow the mantra you are suggesting we would end up defining a separate function that uses `proof` for each and every one of the lenses we've defined.

Reply:
Ok, I added the `EraTallyState era` as a superclass of `Reflect` and got rid of this.